In [ ]:
import pandas as pd
import numpy as np

# Задание 1. Загрузка данных

*Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome (
операция?, возраст, ректальная температура, пульс, частота дыхания, температура конечностей, боль, исход.)*

  1: операция? / surgery?:

          1 = да, была операция
          2 = лечили без операции

  2: Возраст / age:
  
          1 = взрослая лошадь
          2 = Молодой (< 6 месяцев)
          
  4: ректальная температура / temperature of extremities

          - линейный
          - в градусах Цельсия.
          - Повышенная температура может возникнуть из-за инфекции.
          - температура может быть снижена, когда животное находится в позднем шоке
          - нормальная температура 37,8
          - этот параметр обычно меняется по мере развития проблемы
               например. может начаться нормально, затем стать повышенным из-за
                   поражения, проходя обратно через нормальный диапазон, как
                   лошадь в шоке
  5: пульс /pain

          - линейный
          - частота сердечных сокращений в ударах в минуту
          - является отражением состояния сердца: 30-40 нормально для взрослых
          - редко имеют более низкий показатель, чем обычно, хотя спортивные лошади
            может иметь скорость 20-25
          - животные с болезненными поражениями или страдающие циркуляторным шоком
            может быть учащенное сердцебиение

  6: частота дыхания / respiratory rate

          - линейный
          - нормальная скорость от 8 до 10
          - полезность сомнительна из-за больших колебаний

 7: температура конечностей / temperature of extremities

          - субъективное указание на периферическое кровообращение
          - возможные значения:
               1 = нормальный
               2 = теплый
               3 = круто
               4 = холодный
          - холодные или холодные конечности указывают на возможный шок
          - горячие конечности должны коррелировать с повышенной ректальной температурой.

11: боль / pain

          - субъективная оценка уровня боли лошади.
          - возможные значения:
               1 = тревога, боли нет
               2 = депрессия
               3 = перемежающаяся легкая боль
               4 = перемежающаяся сильная боль
               5 = постоянная сильная боль
          - НЕ следует рассматривать как упорядоченную или дискретную переменную!
          - В общем, чем больнее, тем больше вероятность, что потребуется
            операция
          - предшествующее лечение боли может в некоторой степени маскировать уровень боли

 23: исход / outcome

          - что в итоге случилось с лошадью?
          - возможные значения:
               1 = жил
               2 = умер
               3 = был усыплен       

In [ ]:
header = ['surgery?', 'age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temperature_of_extremities', 'peripheral_pulse', 'mucous_membranes', 'capillary_refill_time', 'pain', 'peristalsis', 'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux', 'nasogastric_reflux_PH', 'rectal_examination', 'abdomen', 'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance', 'abdomcentesis_total_protein', 'outcome', 'surgical_lesion?', 'type_of_lesion', 'type_of_lesion_', 'type_of_lesion__', 'cp_data']

In [ ]:
df_initial = pd.read_csv('horse_data.csv', names=header, na_values='?')
df = df_initial.loc[:, ['surgery?', 'age', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temperature_of_extremities','pain', 'outcome']]
df.head()

surgery?  age  rectal_temperature  ...  temperature_of_extremities  pain  outcome
0       2.0    1                38.5  ...                         3.0   5.0      2.0
1       1.0    1                39.2  ...                         NaN   3.0      3.0
2       2.0    1                38.3  ...                         1.0   3.0      1.0
3       1.0    9                39.1  ...                         4.0   2.0      2.0
4       2.0    1                37.3  ...                         NaN   NaN      2.0

[5 rows x 8 columns]

# Задание 2. Первичное изучение данных

*Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.*

**Стобец surgery?**

Значения столбца surgery? могут принимать значения только 1 или 2. Проверим так ли это:

In [ ]:
df['surgery?'].unique()

array([ 2.,  1., nan])

**Стобец age**

Значения столбца age могут принимать значения только 1 или 2. Проверим так ли это:

In [ ]:
df['age'].unique()

array([1, 9])

Вместо значения 2 стоит значение 9. Можно оставить 9 и понимать, что это молодая лошадь, но лучше соответствовать документации и заменить 9 на 2:

In [ ]:
df['age'] = df['age'].replace(9, 2)
df['age'].unique()

array([1, 2])

**Столбец temperature of extremities**

Значения столбца temperature of extremities могут принимать значения только 1, 2, 3, 4. Проверим так ли это:

In [ ]:
df['temperature_of_extremities'].unique()

array([ 3., nan,  1.,  4.,  2.])

**Столбец pain**

Значения столбца pain могут принимать значения только  от 1 до 5. Проверим так ли это:

In [ ]:
df['pain'].unique()

array([ 5.,  3.,  2., nan,  4.,  1.])

**Столбец outcome**

Значения столбца pain могут принимать значения только от 1 до 3. Проверим так ли это:

In [ ]:
df['outcome'].unique()

array([ 2.,  3.,  1., nan])

**Посмотрим информацию о датафрейме:**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


Датафрейм состоит из 300 наблюдений. Работу с пропусками оставим до задания 3.

**Рассчитаем базовые статистики**

In [ ]:
df.describe()

surgery?         age  ...       pain     outcome
count  299.000000  300.000000  ...  245.00000  299.000000
mean     1.397993    1.080000  ...    2.95102    1.551839
std      0.490305    0.271746  ...    1.30794    0.737187
min      1.000000    1.000000  ...    1.00000    1.000000
25%      1.000000    1.000000  ...    2.00000    1.000000
50%      1.000000    1.000000  ...    3.00000    1.000000
75%      2.000000    1.000000  ...    4.00000    2.000000
max      2.000000    2.000000  ...    5.00000    3.000000

[8 rows x 8 columns]

In [ ]:
print('Мода:')
print(f"surgery? - {df['surgery?'].mode()}")
print(f"age - {df['age'].mode()}")
print(f"rectal_temperature	 - {df['rectal_temperature'].mode()}")
print(f"pulse - {df['pulse'].mode()}")
print(f"respiratory_rate - {df['respiratory_rate'].mode()}")
print(f"temperature_of_extremities - {df['temperature_of_extremities'].mode()}")
print(f"pain - {df['pain'].mode()}")
print(f"outcome - {df['outcome'].mode()}")

Мода:
surgery? - 0    1.0
dtype: float64
age - 0    1
dtype: int64
rectal_temperature	 - 0    38.0
dtype: float64
pulse - 0    48.0
dtype: float64
respiratory_rate - 0    20.0
dtype: float64
temperature_of_extremities - 0    3.0
dtype: float64
pain - 0    3.0
dtype: float64
outcome - 0    1.0
dtype: float64


▶  По данной информации можно сделать вывод, что большинство лошадей, испытывающих колики - это взрослые лошади. Большинству лошадей провели операцию. Большинство лошадей выжило. При коликах у лошадей наблюдается повышение температуры, пульса и частоты дыхания. 

**Найдём выбросы**

На выбросы проверим только непрерывные величины - ректальную температуру, пульс и частоту дыхания. Вероятно, здесь стоило рассмотреть и качественные величины (операция, возраст, температура конечностей, боль, исход), но данные столбцы проверены выше (здесь значение 9 в столбце age стало бы выбросом).

При изучении информации в предыдущем пункте бросается в глаза низкие границы пульса и температуры, ведь как указанно в выводе, при коликах эти показатели повышавются. Так же необходимо изучить частоту дыхания, т.к. она имеет очень сильный разброс и отличается от частоты, указанной в документации.

**Столбец rectal_temperature**

In [ ]:
q1 = df['rectal_temperature'].quantile(0.25)
q3 = df['rectal_temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
outliers_rtemperature = df[(-df['rectal_temperature'].between(lower_bound, upper_bound, inclusive=True))&(-df['rectal_temperature'].isnull())].sort_values(by='rectal_temperature')
outliers_rtemperature

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  


surgery?  age  ...  pain  outcome
44        1.0    1  ...   4.0      3.0
141       2.0    1  ...   NaN      2.0
238       2.0    1  ...   3.0      3.0
80        1.0    1  ...   4.0      2.0
118       1.0    1  ...   5.0      1.0
298       1.0    1  ...   3.0      1.0
251       2.0    1  ...   1.0      2.0
99        2.0    1  ...   2.0      1.0
75        1.0    2  ...   2.0      3.0
20        1.0    1  ...   5.0      1.0
281       2.0    1  ...   2.0      2.0
54        2.0    1  ...   2.0      3.0
91        2.0    1  ...   2.0      2.0
259       1.0    1  ...   2.0      2.0

[14 rows x 8 columns]

▶  Случаи, когда температура выше нормальной (37,8), не будем считать за выбросы, т.к. в данных случаях наблюдается повышенный пульс, горячие копыта. В большинстве этих случаев лошади умерли. Вполне логично, что темпертура действительно была такой высокой (кроме того и у людей температура 40 - не выброс, а у лошадей ещё и средняя температура выше человеческой).

▶  Случаи с низкой температурой тоже не будем считать выбросами - они плавно снижаются, а не выбрасываются как кит на берег. Кроме того при температуре ниже 36,5 лошади умерли. Ничего подозрительного, оставляем всё как есть. 

**Столбец pulse**

In [ ]:
q1 = df['pulse'].quantile(0.25)
q3 = df['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
outliers_pulse = df[(-df['pulse'].between(lower_bound, upper_bound, inclusive=True))&(-df['pulse'].isnull())]
outliers_pulse

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  


surgery?  age  ...  pain  outcome
3         1.0    2  ...   2.0      2.0
41        2.0    2  ...   NaN      1.0
55        1.0    2  ...   3.0      2.0
255       1.0    2  ...   4.0      2.0
275       1.0    2  ...   5.0      2.0

[5 rows x 8 columns]

▶  В выбросы попали высокие значения пульса. Но учитывая, что во всех случаях была операция, были боли, и почти все лошади умерли, логично предположить, что у в данных случаях действительно мог быть такой высокий пульс. 

В выброс по данной формеле не попал низкий пульс. Минимум - 30, в документации указано, что нормальный пульс - 30-40. Однако в данном случе у лошадей колики, ожидаемо, что пульс повышен, поэтому посмотрим данные с низким пульсом:

In [ ]:
df.sort_values(by=['pulse']).head(10)

surgery?  age  ...  pain  outcome
232       1.0    1  ...   NaN      1.0
130       1.0    1  ...   NaN      2.0
203       2.0    1  ...   2.0      1.0
276       1.0    1  ...   4.0      3.0
242       1.0    1  ...   5.0      2.0
76        1.0    1  ...   3.0      1.0
166       1.0    1  ...   3.0      1.0
133       2.0    1  ...   2.0      1.0
299       1.0    1  ...   NaN      3.0
112       1.0    1  ...   1.0      1.0

[10 rows x 8 columns]

▶  В двух случаях у лошадей пульс 30. Операция не требовалась, температура не слишком отличается от нормальной. Одна лошадь выжила. Делаем вывод, что это не выброс.

**Столбец respiratory_rate**

In [ ]:
q1 = df['respiratory_rate'].quantile(0.25)
q3 = df['respiratory_rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
outliers_rate = df[(-df['respiratory_rate'].between(lower_bound, upper_bound, inclusive=True))&(-df['respiratory_rate'].isnull())]
outliers_rate

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  


surgery?  age  ...  pain  outcome
3         1.0    2  ...   2.0      2.0
39        1.0    2  ...   NaN      2.0
41        2.0    2  ...   NaN      1.0
82        1.0    2  ...   3.0      1.0
84        1.0    1  ...   2.0      1.0
103       1.0    2  ...   3.0      1.0
106       1.0    1  ...   NaN      1.0
120       1.0    1  ...   2.0      1.0
125       1.0    1  ...   3.0      1.0
186       1.0    1  ...   NaN      1.0
208       1.0    1  ...   NaN      3.0
229       1.0    2  ...   NaN      1.0
244       1.0    2  ...   2.0      1.0
255       1.0    2  ...   4.0      2.0
265       2.0    1  ...   3.0      3.0
269       1.0    1  ...   NaN      1.0
295       1.0    1  ...   2.0      3.0

[17 rows x 8 columns]

In [ ]:
q1 = df['respiratory_rate'].quantile(0.25)
q3 = df['respiratory_rate'].quantile(0.75)
iqr = q3 - q1
iqr

17.5

In [ ]:
q1 = df[-df['respiratory_rate'].isnull()].quantile(0.25)
q3 = df[-df['respiratory_rate'].isnull()].quantile(0.75)
iqr = q3 - q1
iqr

surgery?                       1.0
age                            0.0
rectal_temperature             0.7
pulse                         40.0
respiratory_rate              17.5
temperature_of_extremities     2.0
pain                           2.0
outcome                        1.0
dtype: float64

В данном случае сложно судить, всё ли это выбросы. С одной стороны, нормальное значение по документации 8-10. Но здесь все лошади с коликами, повышенными температурой и пульсом, испытывающие боль. Логично, что ох дыхание учащенно. Кроме того в документации сказано, что польза данных по частоте дыхания сомнительна из-за больших колебаний. Следовательно большие колебания действительно имели место быть. При возможности не использовать данный столбец в дальнейшем анализе.

**Повертим данными перед работой с пропусками, чтобы найти закономерности и т.п.**

 **surgery и age**

Необходимость операции и возраст лошадей - категрии, данные внутри которых могут отличаться. 

In [ ]:
df.pivot_table(index = 'surgery?', columns = 'age', values = ['rectal_temperature', 'pulse', 'respiratory_rate'], aggfunc = ['mean', 'median'], fill_value = 0).head()

mean                                 ...             median                     
              pulse             rectal_temperature  ... rectal_temperature respiratory_rate    
age               1           2                  1  ...                  2                1   2
surgery?                                            ...                                        
1.0       70.254902  131.875000          38.084800  ...               38.6               26  60
2.0       63.405660  100.666667          38.178723  ...               38.6               24  30

[2 rows x 12 columns]

▶  Ректальная температура (средняя и медиана) не сильно отличается в разрезе необходимости операции или возраста (у взрослых на 0,5 выше, чем у жеребят).

▶  Пульс жеребят значительно отличается от пульса взрослых, стоит учесть в задании 3.

In [ ]:
df1_1 = df[(df['surgery?'] == 1.0) & (df['age'] == 1)]
print(df1_1['temperature_of_extremities'].mode())

0    3.0
dtype: float64


In [ ]:
df1_2 = df[(df['surgery?'] == 1.0) & (df['age'] == 2)]
print(df1_2['temperature_of_extremities'].mode())

0    3.0
dtype: float64


In [ ]:
df2_1 = df[(df['surgery?'] == 2.0) & (df['age'] == 1)]
print(df2_1['temperature_of_extremities'].mode())

0    3.0
dtype: float64


In [ ]:
df2_2 = df[(df['surgery?'] == 2.0) & (df['age'] == 2)]
print(df2_2['temperature_of_extremities'].mode())

0    1.0
dtype: float64


▶  Температура конечностей при необходимости операции у взрослых лошадей и у жеребят не отличается (мода). Но при простом лечении у жеребят температура конечностей (мода) отличается от взрослых лошадей. Можно учесть это в задании 3.

Боль вряд ли зависит от возраста лошади. Поэтому её рассмотрим только в разрезе необходимости операции:

In [ ]:
df1 = df[df['surgery?'] == 1.0]
df1['pain'].mode()


0    3.0
dtype: float64

In [ ]:
df2 = df[df['surgery?'] == 2.0]
df2['pain'].mode()

0    1.0
dtype: float64

▶  При необходимоти операции уровень боли был выше, чем при простом лечении. Учесть в заднии 3.

**rectal_temperature vs temperature_of_extremities**

Есть ли взаимосвязь между температурой ректальной и температурой конечностей

In [ ]:
df.pivot_table(index = 'temperature_of_extremities', columns = 'surgery?', values = 'rectal_temperature', aggfunc = ['mean', 'median'], fill_value = 0).head()

mean            median      
surgery?                          1.0        2.0    1.0   2.0
temperature_of_extremities                                   
1.0                         38.129268  38.122581   38.1  38.2
2.0                         38.155556  38.229412   38.0  38.2
3.0                         38.095652  38.321212   38.1  38.3
4.0                         38.553333  38.160000   38.5  38.2

▶  Взаимосвязи между этими двумя температурами почти нет.

**rectal_temperature vs pulse**

Будет ли подниматься пульс при повышении температуры? Коэффициент корреляции говорит "нет", взаимосвязмь между этими показателями слабая.

In [ ]:
df['rectal_temperature'].corr(df['pulse'])

0.2192601672244764

На всякий случай проверим взаимосвязь и с температурой конечностей

In [ ]:
df['temperature_of_extremities'].corr(df['pulse'])

0.3350410084776435

# Задание 3. Работа с пропусками

*Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них.
 Сформировать датафрейм, в котором пропуски будут отсутствовать.*

Расчитаем процентное соотношение пропущенных значений для каждого столбца:

In [ ]:
(df.isna().mean() * 100).round(2).sort_values()

age                            0.00
surgery?                       0.33
outcome                        0.33
pulse                          8.00
pain                          18.33
temperature_of_extremities    18.67
respiratory_rate              19.33
rectal_temperature            20.00
dtype: float64

**surgery?**

In [ ]:
df[df['surgery?'].isnull()]

surgery?  age  ...  pain  outcome
132       NaN    1  ...   4.0      NaN

[1 rows x 8 columns]

In [ ]:
df.dropna(subset = ['surgery?'], inplace = True)

▶  В данной строке нельзя восстановить данные по столбцам surgery? и outcome. Однако это важные показатели, без которых наблюдение не имеет смысла. Удалим эту строку. Т.к. это единичный случай, когда не тданных по surgery? и outcome, на конечный результат удаление строки не повлияет.

In [ ]:
df[(df['rectal_temperature'].isnull())&(df['pulse'].isnull())&(df['respiratory_rate'].isnull())&(df['temperature_of_extremities'].isnull())&(df['pain'].isnull())]

surgery?  age  ...  pain  outcome
28        1.0    1  ...   NaN      2.0
56        1.0    1  ...   NaN      1.0
74        1.0    2  ...   NaN      2.0
175       1.0    1  ...   NaN      1.0
207       1.0    1  ...   NaN      2.0
216       2.0    1  ...   NaN      1.0
227       1.0    2  ...   NaN      1.0
288       1.0    1  ...   NaN      1.0

[8 rows x 8 columns]

▶  Данные строки нельзя принять к анализу и заменить нельзя, т.к. это будут абсолютно новые наблюдения. Удаляем. 

In [ ]:
df.drop(index=[28, 56, 74, 175, 207,  216, 227, 288], inplace = True)

In [ ]:
(df.isna().mean() * 100).round(2).sort_values()

surgery?                       0.00
age                            0.00
outcome                        0.00
pulse                          5.50
pain                          16.15
temperature_of_extremities    16.49
respiratory_rate              17.18
rectal_temperature            17.87
dtype: float64

**pulse**

Пропусков в столбце pulse 5.5%. Разброс значений довольно большой. Так же в наблюдениях, где пропуски пульса, нет и других параметров. Поэтому удалим данные строки. Восстановим только те, где ректальная температура известна (c учётом выводов из задания 2). 

In [ ]:
df[df['pulse'].isnull()]

surgery?  age  ...  pain  outcome
5         2.0    1  ...   2.0      1.0
52        2.0    1  ...   1.0      1.0
58        1.0    1  ...   5.0      2.0
78        1.0    1  ...   5.0      2.0
83        1.0    1  ...   5.0      2.0
93        2.0    1  ...   5.0      2.0
115       2.0    1  ...   3.0      1.0
117       2.0    1  ...   3.0      3.0
126       1.0    1  ...   5.0      1.0
136       1.0    1  ...   5.0      2.0
151       1.0    1  ...   4.0      2.0
159       2.0    1  ...   2.0      1.0
160       1.0    1  ...   4.0      1.0
174       2.0    1  ...   4.0      2.0
182       1.0    1  ...   5.0      2.0
204       1.0    1  ...   4.0      1.0

[16 rows x 8 columns]

In [ ]:
df = df.loc[~((df['pulse'].isnull())&(df['rectal_temperature'].isnull()))]

In [ ]:
df['pulse'] = df['pulse'].fillna(df.groupby(['surgery?', 'age'])['pulse'].transform('median'))

**temperature_of_extremities**

Столбцы temperature_of_extremities и pain	- качественные, поэтому заменим пропуски их модой.

In [ ]:
df['temperature_of_extremities'] = df['temperature_of_extremities'].fillna(df['temperature_of_extremities'].mode()[0])

**pain**

Логично, что в случае, когда лошадь нуждалась в операции, боль была сильнее, чем в случае, когда операция не требовалась. Поэтому пропуски заменим модой в разрезе значений  столбца surgery?

In [ ]:
df.loc[(df['surgery?'] == 1.0)&(df['pain'].isnull()), 'pain'] = 3.0
df.loc[(df['surgery?'] == 2.0)&(df['pain'].isnull()), 'pain'] = 1.0


In [ ]:
(df.isna().mean() * 100).round(2).sort_values()

surgery?                       0.00
age                            0.00
pulse                          0.00
temperature_of_extremities     0.00
pain                           0.00
outcome                        0.00
rectal_temperature            14.95
respiratory_rate              15.30
dtype: float64

**rectal_temperature**

Пропуски в столбце rectal_temperature заменим на среднее значение: разброс небольшой (стандартное отклонение = 0,7), среднее и медиана отличаются лишь на 0,1. Учтём разницу температуры у взрослых лошадей и жеребят.

In [ ]:
df['rectal_temperature'] = df['rectal_temperature'].fillna(df.groupby('age')['rectal_temperature'].transform('mean'))

In [ ]:
(df.isna().mean() * 100).round(2).sort_values()

surgery?                       0.0
age                            0.0
rectal_temperature             0.0
pulse                          0.0
temperature_of_extremities     0.0
pain                           0.0
outcome                        0.0
respiratory_rate              15.3
dtype: float64

**respiratory_rate**

Остался столбец respiratory_rate. Это очень спроный столбец, с одной стороны, хорошо бы его восстановить, но с другой, в этом столбце значения очень разбросаны, и в самой документации сказано, что из-за высоких колебаний польза данного столбца сомнтиельна. Удалим этот столбец и будем уповать на то, что он не понадобится. Если нужен - откатим. Но в реально жизни целесообразность данного столбца необходимо бы было обсудить до началапредобработки данных. А так можно ещё и в формуле потренироваться =)

In [ ]:
df_final = df.drop(['respiratory_rate'], axis=1)
df_final.head()

surgery?  age  rectal_temperature  ...  temperature_of_extremities  pain  outcome
0       2.0    1                38.5  ...                         3.0   5.0      2.0
1       1.0    1                39.2  ...                         3.0   3.0      3.0
2       2.0    1                38.3  ...                         1.0   3.0      1.0
3       1.0    2                39.1  ...                         4.0   2.0      2.0
4       2.0    1                37.3  ...                         3.0   1.0      2.0

[5 rows x 7 columns]

**дубликаты**

Как финальный штрих проверим дубликаты

In [ ]:
df_final.duplicated().sum()

15

In [ ]:
df_final[df_final.duplicated(keep=False)]

surgery?  age  ...  pain  outcome
26        1.0    1  ...   3.0      1.0
34        1.0    1  ...   5.0      1.0
38        2.0    1  ...   3.0      1.0
40        1.0    1  ...   5.0      2.0
48        1.0    1  ...   2.0      2.0
59        2.0    1  ...   5.0      2.0
67        2.0    1  ...   1.0      1.0
71        1.0    1  ...   1.0      1.0
73        1.0    1  ...   4.0      2.0
83        1.0    1  ...   5.0      2.0
86        1.0    1  ...   5.0      2.0
94        2.0    1  ...   1.0      1.0
110       1.0    1  ...   2.0      2.0
143       1.0    1  ...   1.0      1.0
146       1.0    1  ...   3.0      1.0
156       2.0    1  ...   3.0      1.0
179       2.0    1  ...   4.0      3.0
182       1.0    1  ...   5.0      2.0
186       1.0    1  ...   3.0      1.0
192       1.0    1  ...   3.0      1.0
205       2.0    1  ...   4.0      3.0
225       2.0    1  ...   1.0      1.0
233       1.0    1  ...   3.0      1.0
234       2.0    1  ...   1.0      1.0
235       1.0    1  ...   4.0      2.0
245       2.0    1  ...   5.0      2.0
269       1.0    1  ...   3.0      1.0
272       1.0    1  ...   5.0      1.0
282       1.0    1  ...   2.0      2.0
291       1.0    1  ...   2.0      2.0

[30 rows x 7 columns]

В данных нет идентификатора лошадей и есть вероятность, что у двух лошадей полностью совпали данные. Но температура лошадей в некоторых случаях совпадает вплоть до тысячного знака после запятой. Сделаем вывод, что это дубликаты и удалим их.

In [ ]:
df_final = df_final.drop_duplicates().reset_index(drop= True)

In [ ]:
df_final

surgery?  age  ...  pain  outcome
0         2.0    1  ...   5.0      2.0
1         1.0    1  ...   3.0      3.0
2         2.0    1  ...   3.0      1.0
3         1.0    2  ...   2.0      2.0
4         2.0    1  ...   1.0      2.0
..        ...  ...  ...   ...      ...
261       1.0    1  ...   2.0      3.0
262       2.0    1  ...   4.0      3.0
263       1.0    1  ...   4.0      2.0
264       1.0    1  ...   3.0      1.0
265       1.0    1  ...   3.0      3.0

[266 rows x 7 columns]

In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    266 non-null    float64
 1   age                         266 non-null    int64  
 2   rectal_temperature          266 non-null    float64
 3   pulse                       266 non-null    float64
 4   temperature_of_extremities  266 non-null    float64
 5   pain                        266 non-null    float64
 6   outcome                     266 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 14.7 KB


Такми образом к дальнейшей работе готов датафрейм, состоящий из 7 колонок и 266 строк без пропусков и дубликатов.